### PR TITLE
Implemented custom duplicate key handling (#840)

### PR DESCRIPTION
**Pull request by XIAYM-gh**
Supports: throw a JSONException (by default), ~ignore (not overwritting)~, overwrite ~or merge into a JSONArray~;
also with tests in a separated file

xD, thanks for reviewing

### CODE REVIEW

**Comment by XIAYM-gh** [comment]
whoa, 2 different errors, working on them

**Comment by XIAYM-gh** [comment]
I can't believe it throws a StackOverFlowError while compiling, should I add a -Xss4M jvm argument to the actions file?

**Comment by stleary** [comment]
> I can't believe it throws a StackOverFlowError while compiling, should I add a -Xss4M jvm argument to the actions file?

It appears to be an intermittent error, only seems to affect Java v11, root cause unknown. After a couple of retries it was successful. Also, the build and test is successful on my laptop. No changes needed at this time.

However, your code is out of date, please merge from the latest master branch.

File: src/main/java/org/json/JSONParserConfiguration.java [hunk 1]
```diff
@@ -4,23 +4,49 @@
...
+        return new JSONParserConfiguration();
+    }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    public JSONParserConfiguration withMaxNestingDepth(final int maxNestingDepth) {
+        return super.withMaxNestingDepth(maxNestingDepth);
+    }
+
+    public JSONParserConfiguration withOverwriteDuplicateKey(final boolean overwriteDuplicateKey) {
```
Comment (stleary) [inline src/main/java/org/json/JSONParserConfiguration.java:42]: Please include JavaDocs for this and the next method.
image

File: src/test/java/org/json/junit/JSONObjectDuplicateKeyTest.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package org.json.junit;
+
+import org.json.*;
+
+import static org.junit.Assert.*;
+
+import org.junit.Test;
+
```
Comment (stleary) [inline src/test/java/org/json/junit/JSONObjectDuplicateKeyTest.java:8]: Please include these unit tests:
```
    @Test
    public void verifyDuplicateKeyThenMaxDepth() {
        JSONParserConfiguration jsonParserConfiguration = new JSONParserConfiguration()
                .withOverwriteDuplicateKey(true)
                .withMaxNestingDepth(42)
                ;

        assertEquals(42, jsonParserConfiguration.getMaxNestingDepth());
        assertEquals(true, jsonParserConfiguration.isOverwriteDuplicateKey());
    }

    @Test
    public void verifyMaxDepthThenDuplicateKey() {
        JSONParserConfiguration jsonParserConfiguration = new JSONParserConfiguration()
                .withMaxNestingDepth(42)
                .withOverwriteDuplicateKey(true)
                ;

         assertEquals(true, jsonParserConfiguration.isOverwriteDuplicateKey());
         assertEquals(42, jsonParserConfiguration.getMaxNestingDepth());
    }
```

File: src/test/java/org/json/junit/JSONObjectDuplicateKeyTest.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package org.json.junit;
```
Comment (stleary) [inline src/test/java/org/json/junit/JSONObjectDuplicateKeyTest.java:1]: Please rename to JSONParserConfigurationTest

**Comment by XIAYM-gh** [comment]
I found the isKeepStrings() method is unused, should I overwrite the method and throws an UnsupportedOperationException?

**Comment by stleary** [comment]
> I found the isKeepStrings() method is unused in the JSONParserConfiguration class, should I overwrite the method and throws an UnsupportedOperationException?

That method is used by JSONML and XML classes, please don't make any changes to it.

**Comment by XIAYM-gh** [comment]
I see, thanks!

**Comment by stleary** [comment]
**What problem does this code solve?**
Adds an opt-in ability to overwrite a JSONObject duplicate key instead of throwing an exception
 
**Does the code still compile with Java6?**
Yes

**Risks**
Low

**Changes to the API?**
Yes, new parsing options were added

**Will this require a new release?**
No

**Should the documentation be updated?**
No

**Does it break the unit tests?**
No, new unit tests were added

**Was any code refactored in this commit?**
No

**Review status**
**APPROVED**

Starting 3-day comment window

**Comment by stleary** [comment]
@XIAYM-gh Please merge from the latest master branch, the PR is not up to date.

**Comment by XIAYM-gh** [comment]
Oh, I apologize for the late reply, I was just back from school, and I'm sorry for the inconvenience that I've caused

**Review by stleary — APPROVED** [review]
The new JSONParserConfiguration ctor will be removed in a followup commit.

**Pull request merged**